src/main/java/org/mutabilitydetector/checkers/ImmutableCollectionChecker.java

```diff
@@ -177,41 +179,28 @@ private String typeAssignedToField() {
         return CLASS_NAME_CONVERTER.dotted(typeAssignedToField.getInternalName());
     }
 
-    private boolean cannotGetMethodNameAndOnClassFromPreviousInvocation(String[] retVal) {
+    private Optional<MethodInsnNode> getLastMethodInsnNode() {
         AbstractInsnNode lastMeaningfulNode = lastMeaningfulNode(fieldInsnNode);
         if (!(lastMeaningfulNode instanceof MethodInsnNode)) {
-            return true;
+            return Optional.of((MethodInsnNode) lastMeaningfulNode);
+        } else {
+            return Optional.empty();
         }
-
-        MethodInsnNode previousInvocation = (MethodInsnNode) lastMeaningfulNode;
-        retVal[0] = previousInvocation.name;
-        retVal[1] = previousInvocation.owner;
-
-        return false;
     }
 
     public boolean checkInvokesImmutableInterfaceMethod() {
-        String[] methodNameAndOnClassAry = new String[2];
-        if (cannotGetMethodNameAndOnClassFromPreviousInvocation(methodNameAndOnClassAry)) {
-            return false;
-        }
-        String methodName = methodNameAndOnClassAry[0];
-        String onClass = methodNameAndOnClassAry[1];
+        return getLastMethodInsnNode().map(previousInvocation ->
+                        Configuration.INSTANCE.FIELD_TYPE_TO_IMMUTABLE_METHOD.containsKey(CLASS_NAME_CONVERTER.dotted(previousInvocation.name)) &&
+                        Configuration.INSTANCE.FIELD_TYPE_TO_IMMUTABLE_METHOD.get(typeAssignedToField()).equals(previousInvocation.owner))
+                    .orElse(false);
 
-        return Configuration.INSTANCE.FIELD_TYPE_TO_IMMUTABLE_METHOD.containsKey(CLASS_NAME_CONVERTER.dotted(onClass))
-                && Configuration.INSTANCE.FIELD_TYPE_TO_IMMUTABLE_METHOD.get(typeAssignedToField()).equals(methodName);
     }
 
     private boolean wrapsInUnmodifiable() {
-        String[] methodNameAndOnClassAry = new String[2];
-        if (cannotGetMethodNameAndOnClassFromPreviousInvocation(methodNameAndOnClassAry)) {
-            return false;
-        }
-        String methodName = methodNameAndOnClassAry[0];
-        String onClass = methodNameAndOnClassAry[1];
-
-        return Configuration.INSTANCE.UNMODIFIABLE_METHOD_OWNER.equals(CLASS_NAME_CONVERTER.dotted(onClass))
-                && Configuration.INSTANCE.FIELD_TYPE_TO_UNMODIFIABLE_METHOD.get(typeAssignedToField()).equals(methodName);
+        return getLastMethodInsnNode().map(previousInvocation ->
+                        Configuration.INSTANCE.UNMODIFIABLE_METHOD_OWNER.equals(CLASS_NAME_CONVERTER.dotted(previousInvocation.owner)) &&
+                        Configuration.INSTANCE.FIELD_TYPE_TO_UNMODIFIABLE_METHOD.get(typeAssignedToField()).equals(previousInvocation.name))
+                    .orElse(false);
```

Comment:
So I know you wanted me to use a method reference. I've mainly used them when IntelliJ has suggested them for replacing a lambda. I can't use it here because of both `onClass` and `methodName` being required for the boolean expression.

Hear me out here. What I might be able to do is declare something like this:
`private static final Function<MethodInsnNode, Boolean> someFuncName = previousInvocation -> ...` And then I could just supply that function to `Optional#map()` in both `checkInvokesImmutableFactoryMethod()` and `wrapsInUmodifiable()`.

I am thinking that what we could do is merge `FIELD_TYPE_TO_UNMODIFIABLE_METHOD` and `FIELD_TYPE_TO_IMMUTABLE_METHOD`. I'm also thinking maybe I can add a dummy key-value pair to the merged map with `java.util.Collections` as the key in that dummy key-value pair. Because for the Java 9 Immutable factory methods, the `owner` depends on whether it's a `List`, a `Set` or a `Map`. Hmm, do you see any problems with using `containsKey()` and that dummy KV pair for the unmodifiable collections?

I'm only suggesting this because you suggested a method reference. In order to supply one lambda for both methods, there needs to be generalization between the two methods.

![unmodifiableList](https://user-images.githubusercontent.com/15847948/86194794-71728e00-bb1d-11ea-93cd-bd3900060bbc.png)

Reply:
Actually thinking about this, if I could generalize then `wrapInUnmodifiable()` and `checkInvokesImmutableFactoryMethod()` could be consolidated into one method. Also thinking about this, a lambda function wouldn't be the best option because the `Boolean` result is based on the state of the instance.

Reply:
But maybe it's better to leave them as 2 methods. Because there's more checking after `wrapsInUnmodifiable()` is finally called.

Reply:
@fma1 what you have now looks good to me. I had in mind 2x public methods + 2x private methods used as a method reference (of type `Function<MethodInsnNode, Boolean>` as you say). But I think 2x public methods, each with an inline lambda, is better.